components/prism-hcl.js

```diff
@@ -0,0 +1,95 @@
+Prism.languages.hcl = {
+	'comment': [
+		{
+			pattern: /(^|[^\\])\/\*[\s\S]*?(?:\*\/|$)/,
+			lookbehind: true
+		},
+		{
+			pattern: /(^|[^\\:])\/\/.*/,
+			lookbehind: true,
+			greedy: true
+		},
+		{
+			pattern: /#.*/
+		}
+	],
+	'keyword': [
+		{
+			pattern: /(resource|data)\s+(")?(\w+)(")?\s+(?=(")?([\w\-]+)(")?\s+{)/i,
+			inside: {
+				'type': [
+					{
+						pattern: /(resource|data)\s+(")?(\w+)(")?\s+/i,
+						lookbehind: true,
+						alias: 'variable'
+					}
+                ]
+			}
+		},
+		{
+			pattern: /(provider|provisioner|variable|output|module|backend)\s+(")?([\w\-]+)(")?\s+(?={)/i,
+			inside: {
+				'type': [
+					{
+						pattern: /(provider|provisioner|variable|output|module|backend)\s+(")?([\w\-]+)(")?\s+/i,
+						lookbehind: true,
+						alias: 'variable'
+					}
+                ]
+			}
+		},
+		{
+			pattern: /([\w\-]+)(?=\s+{)/i
+		}
+	],
+	'interpolation': [
+		{
+			pattern: /\${([\w\[\]\(\+\)\-\*\%\^\"\'\,\?\=\:\/\s]+\.?)+}/i,
+			inside: {
+				'type': [
+					{
+						pattern: /((terraform|var|self|count|module|path|data|local)\.)[\w\*]+/i,
+						lookbehind: true,
+						alias: 'variable'
+					}
+				],
+				'keyword': [
+					{
+						pattern: /(terraform|var|self|count|module|path|data|local)/i
+					}
+				],
+				'operator': [
+					{
+						pattern: /\${/
+					},
+					{
+						pattern: /}/
+					},
+					{
+						pattern: /\./
+					}
+				],
+				'function': [
+					{
+						pattern: /[\w]+(?=\()/i,
+						alias: 'variable'
+					}
+				],
+				'string': /".*"/i,
+				'punctuation': /[!"#%&'()*+,.\/;<=>@\[\\\]^`{|}~]/,
+				'number': /-?\d+\.?\d*/,
+
+				'others': [
+					{
+						pattern: /[\w\*]+/i,
+						alias: 'string'
+					}
+				],
+			},
+		}
+	],
+	'string': /(["'])(\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*\1/,
```

Comment:
2. I think greedy is not needed here.

Reply:
2. Comments alone can destroy your non-greedy strings: `"abc // what now?"`

Reply:
1. HCL support only double-quoted strings.

2. It is hard to handle. HCL heavily use double-quoted strings like below. That's why I said greedy is not needed. If string is greedy, all interpolation in double-quoted is treated as string. Any idea?

![2018-11-21 02 43 44](https://user-images.githubusercontent.com/390146/48792375-7400e500-ed37-11e8-8474-079cac1db597.png)

Reply:
Just move `interpolation` into `['string'].inside`. I mean, isn't inpolation just part of a string?

This necessitates that the `string` pattern itself is aware of interpolation and thus it has to be adjusted accordingly.
Question: Does HCL support nested interpolation e.g. `"${file("/${var.trouble}.txt")}"` ?

Reply:
> Just move interpolation into ['string'].inside. I mean, isn't inpolation just part of a string?

This is a good approach. I will try to find a way to support `interpolation` under `string`. I didn't fully understand prism's language definitions.

Reply:
Maybe a little help from my side: The `string` pattern should be:

```js
/"(?:[^\n\r\\$"]|\\.|\$(?:(?="|\$(?!\{))|[^\n\r\{$]|\$\{|\{(?:[^\n\r}"]|"(?:[^\n\r\\"]|\\.)*")*\}))*"/
```

It's more complex because of [this](https://www.terraform.io/docs/configuration/interpolation.html):
> You can escape interpolation with double dollar signs: `$${foo}` will be rendered as a literal `${foo}`.

Reply:
I tried it. However, I feels it is too complex to maintain to me.

Reply:
@outsideris 
I'm going to reuse this thread when talking about interpolation. I hope that's alright.

---

This pattern is a little simpler (and is based on the new string pattern): 
_(1)_

```js
/"(?:[^\\$"]|\\[\s\S]|\$(?:(?=")|\$+|[^"${])|\$\{(?:[^{}"]|"(?:[^\\"]|\\[\s\S])*")*\})*"/
```

I can think of simpler patterns but all of them will fail in O(2^n) in certain cases, meaning that they are practically useless.
So that is the simplest (useful) pattern I can think of which handles the following cases correctly:

```hcl
"foo\nbar" // simple

"$foo", "a$b", "foo$" // strings containing $
"${","}" // simple interpolation

// two strings in each line because the interpolation is escaped
// (GitHub gets all of these wrong)
"$${","}"
"$$${","}"
"$$$${","}"
"$$$$${","}"
```

We can then use the following pattern to get the interpolations parts of a string: 
_(2)_

```js
{
    pattern: /(^|[^$])\$\{(?:[^{}"]|"(?:[^\\"]|\\[\s\S])*")*\}/,
    lookbehind: true
}
```
(Everything after the lookbehind is copied from the above string pattern _(1)_.)

Using these two patterns we can make a string pattern similar to [JavaScript's template strings](https://github.com/PrismJS/prism/blob/master/components/prism-javascript.js#L60) using _(1)_ for `template-string` and _(2)_ for `interpolation`.
We have to adjust the inside of `interpolation` of course, but for that, we can use your current interpolation's inside.